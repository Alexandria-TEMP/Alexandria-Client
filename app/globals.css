@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}

@layer base {
  h1 {
    @apply text-2xl font-bold;
  }
  h2 {
    @apply text-xl font-semibold;
  }
  h3 {
    @apply text-lg;
  }
  h6 {
    @apply text-sm;
  }
}
