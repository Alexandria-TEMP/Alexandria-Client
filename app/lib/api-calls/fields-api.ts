import useSWR, { SWRResponse } from "swr";
import { ScientificFieldT } from "../types/api-types";
import { baseUrl, validateResponse } from "./api-common";

/**
 * Hook that fetches all scientific fields from the database, using SWR for caching, loading and error states
 * Fetching all tags from the database sounds like a very bad idea, but we would need some kind of lazy loading for MultiSelectAutocomplete otherwise
 * that is currently in todo
 * @returns the scientific field objects as array, as well as other SWR states (loading, error)
 */
export function useScientificFields(): SWRResponse<ScientificFieldT[], Error> {
  return useSWR(baseUrl + "/tags/scientific", async (...args) => {
    const response = await fetch(...args);
    await validateResponse(response);
    //disable reason: idk how to fix this and still get the correct type cause typescript
    // i have to look into this
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const scientificFieldTags: ScientificFieldT[] = await response.json();
    return scientificFieldTags;
  });
}
