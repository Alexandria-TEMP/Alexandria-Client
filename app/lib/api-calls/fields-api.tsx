// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

export async function getFields() {
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}
