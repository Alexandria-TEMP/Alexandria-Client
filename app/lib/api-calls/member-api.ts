// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

import useSWR, { SWRResponse } from "swr";
import { MemberT, idT } from "../types/api-types";
import { baseUrl, validateResponse } from "./api-common";

/**
 * Gets data for a Member given their ID.
 * @async
 * @param id Member ID
 */
export default async function getMemberData(id: idT): Promise<MemberT> {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));
  if (id == 1) {
    return {
      id,
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
      scientificFields: [],
    };
  } else if (id == 2) {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      lastName: "Copernicus",
      scientificFields: [],
    };
  } else if (id == 3) {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Metal Bar",
      institution: "TU Delft",
      lastName: "Clanging",
      scientificFields: [],
    };
  } else {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Michael",
      institution: "TU Delft",
      lastName: "Yippie",
      scientificFields: [],
    };
  }
}

/**
 * Hook that fetches all members from the database, using SWR for caching, loading and error states
 * Fetching all members from the database sounds like a very bad idea, but we would need some kind of lazy loading for MultiSelectAutocomplete otherwise
 * that is currently in todo
 * NOTE: currently the endpoint actually only returns id, first and last names as that is all that is needed
 * @returns the member objects as array, as well as other SWR states (loading, error)
 */
export function useFetchMembers(): SWRResponse<MemberT[], Error> {
  return useSWR(baseUrl + "/members", async (...args) => {
    const response = await fetch(...args); //by default a get request
    await validateResponse(response);
    const members: MemberT[] = (await response.json()) as MemberT[];
    return members;
  });
}
