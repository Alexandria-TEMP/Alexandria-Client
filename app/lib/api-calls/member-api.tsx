// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

/**
 * Gets data for a Member given their ID.
 * @async
 * @param id Member ID
 */
export default async function getMemberData(id: string) {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));
  if (id == "1") {
    return {
      id: "1",
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      picture: "/placeholders/Marie_Curie.jpg",
      institution: "TU Delft",
      lastName: "Curie",
    };
  } else {
    return {
      id: "2",
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Copernicus",
    };
  }
}

/**
 * Method that gets all members from the database
 * @returns A map of all members with their database id as key, and the whole member object as value
 */
export function getMembers() {
  // TODO

  // simulate long time to get list of users, this should be done without fetching all users
  // ideally we would fetch as we go but i dont really know how to do that
  const members = [
    {
      id: "1",
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      picture: "/placeholders/Marie_Curie.jpg",
      institution: "TU Delft",
      lastName: "Curie",
    },
    {
      id: "2",
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Copernicus",
    },
    {
      id: "3",
      email: "kopernicus@tudelft.nl",
      firstName: "Metal Bar",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Clanging",
    },
    {
      id: "4",
      email: "kopernicus@tudelft.nl",
      firstName: "Michael",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Yippie",
    },
  ];

  return new Map(members.map((m) => [m.id, m]));
}
