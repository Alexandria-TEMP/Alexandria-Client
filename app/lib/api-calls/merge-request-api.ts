import { BranchT, idType } from "../types/api-types";

/**
 * Gets data for a branch given their ID.
 * @async
 * @param id branch ID
 */
export async function getBranchData(id: idType): Promise<BranchT> {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 800));
  if (id == 1)
    return {
      id: id,
      newPostTitle: "Post title",
      projectPostID: 1,
      branchTitle: "Remove contents section",
      newVersionID: 1,
      reviewIDs: [3, 1, 2],
      anonymous: false,
      createdAt: "19 May 2024",
      collaboratorIDs: [0, 1],
      updatedAt: "20 May 2024",
      updatedCompletionStatus: "ideation",
      updatedScientificFields: ["Mathematics"],
      branchReviewStatus: "rejected",
      previousVersionID: 2,
    };
  else if (id == 2)
    return {
      id: id,
      newPostTitle: "Post title",
      projectPostID: 1,
      branchTitle: "Do some stuff",
      newVersionID: 1,
      reviewIDs: [3, 2],
      anonymous: false,
      createdAt: "19 May 2024",
      collaboratorIDs: [0, 1],
      updatedAt: "20 May 2024",
      updatedCompletionStatus: "ideation",
      updatedScientificFields: ["Mathematics"],
      branchReviewStatus: "open for review",
      previousVersionID: 2,
    };
  else
    return {
      id: id,
      newPostTitle: "Post title",
      projectPostID: 1,
      branchTitle: "Grammar fixes",
      newVersionID: 1,
      reviewIDs: [2, 4, 3],
      anonymous: false,
      createdAt: "19 May 2024",
      collaboratorIDs: [0, 1],
      updatedAt: "20 May 2024",
      updatedCompletionStatus: "ideation",
      updatedScientificFields: ["Mathematics"],
      branchReviewStatus: "peer reviewed",
      previousVersionID: 2,
    };
}

/**
 * Gets branches of a post given their ID.
 * @async
 * @param id Post ID
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getPostBranches(id: idType) {
  await new Promise((resolve) => setTimeout(resolve, 100));
  return {
    rejected: ["1", "1", "1", "1"],
    open: ["2", "3", "3"],
    accepted: [
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
      "4",
    ],
  };
}

/**
 * TODO jsdoc when properly implemented
 */
export async function getBranchReviewStatuses(id: idType) {
  await new Promise((resolve) => setTimeout(resolve, 70));
  if (id == 1) return ["accept", "reject", "accept"];
  else if (id == 2) return ["accept", "open", "open"];
  else return ["accept", "accept", "accept"];
}
