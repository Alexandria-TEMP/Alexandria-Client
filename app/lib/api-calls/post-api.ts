import { PostT, PostCreationFormT, idT } from "../types/api-types";
import { validateResponse } from "./api-common";
import { baseUrl } from "./api-common";

/**
 * Gets data for a Post given their ID.
 * @async
 * @param id Post ID
 */
export default async function getPostData(id: idT): Promise<PostT> {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));

  return {
    title: "Post title",
    discussionIDs: [],
    renderStatus: "failure",
    collaboratorIDs: [1, 2], /// i think these should be different, but branch and post are strutured differently so we need both
    id: id,
    postType: "reflection",
    scientificFields: ["1", "2", "3"],
  };
}

/**
 * Method that sends a POST request to the server to create a new post, currently only metadata
 * @async
 * @param postCreationForm object containing post creation form data
 * @returns the newly created post
 */
export async function postPosts(
  postCreationForm: PostCreationFormT,
): Promise<PostT> {
  const jsonPost = JSON.stringify(postCreationForm);
  const response = await fetch(baseUrl + "/posts", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: jsonPost,
  });
  await validateResponse(response);
  //disable reason: idk how to fix this and still get the correct type cause typescript
  // i have to look into this
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const post: PostT = await response.json();
  return post;
}

/**
 * Method that sends a POST request to the server to upload a new file to an existing (non project) post
 * File technically gets sent as multipart form data
 * @async
 * @param postId the id of the post we want to upload files to
 * @param file the file object
 * @returns whether the request retuned a 200OK response
 */
export async function postPostsIdUpload(postId: idT, file: File) {
  const fileData = new FormData();
  fileData.append("file", file);

  const response = await fetch(baseUrl + "/posts/" + postId + "/upload", {
    method: "POST",
    body: fileData,
  });
  await validateResponse(response);
  return response.ok;
}
