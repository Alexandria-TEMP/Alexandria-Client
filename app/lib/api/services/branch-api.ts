import { BranchUnionT, idBranchUnionT } from "@/lib/types/branch-union";
import { BranchT, ClosedBranchT, idT } from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";
import fetchPostData from "./post-api";

/**
 * Fetches branch or closed branch data in a unified object
 * @async
 * @param id branch or closed branch ID
 * @returns branch and optionally closed branch data
 */
export async function fetchBranchData(
  id: idBranchUnionT,
): Promise<BranchUnionT> {
  let closedBranch = undefined;

  if (id.isClosed) {
    const closedBranchResponse = await fetch(
      `${baseUrl}/branches/closed/${id.id}`,
    );
    await validateResponse(closedBranchResponse);
    closedBranch = (await closedBranchResponse.json()) as ClosedBranchT;
  }

  const branchID = closedBranch?.branchID ?? (id.id as idT);
  const branchResponse = await fetch(`${baseUrl}/branches/${branchID}`);
  await validateResponse(branchResponse);

  const branch = (await branchResponse.json()) as BranchT;

  const projectPostID = branch.projectPostID ?? closedBranch?.projectPostID;
  if (!projectPostID) {
    throw new Error(
      `missing some project post ID in ${id.isClosed ? "closed" : ""} branch ${id.id}`,
    );
  }

  const updated = await fetchBranchUpdatedFieldsFallback(branch, projectPostID);

  return { branch, closedBranch, updated, projectPostID };
}

/**
 * For each of the possible post fields that a branch updates, returns
 * either the updated data or the current data if updated is null
 * @param branch branch whose update we're interested in
 * @param projectPostID ID of project post that branch is updating
 */
export async function fetchBranchUpdatedFieldsFallback(
  branch: BranchT,
  projectPostID: idT,
) {
  if (
    branch.updatedPostTitle &&
    branch.updatedCompletionStatus &&
    branch.updatedScientificFieldTagContainerID
  ) {
    return {
      postTitle: branch.updatedPostTitle,
      completionStatus: branch.updatedCompletionStatus,
      scientificFieldTagContainerID:
        branch.updatedScientificFieldTagContainerID,
    };
  }

  const postData = await fetchPostData({
    id: projectPostID,
    isProject: true,
  });

  return {
    postTitle: branch.updatedPostTitle ?? postData.post.title,
    completionStatus:
      branch.updatedCompletionStatus ??
      postData.projectPost!.projectCompletionStatus,
    scientificFieldTagContainerID:
      branch.updatedScientificFieldTagContainerID ??
      postData.post.scientificFieldTagContainerID,
  };
}

/**
 * Gets branches of a post given their ID.
 * @async
 * @param id Post ID
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getPostBranches(id: idT) {
  await new Promise((resolve) => setTimeout(resolve, 100));
  // TODO
  return {
    rejected: [1],
    open: [2],
    accepted: [3],
  };
}

/**
 * TODO jsdoc when properly implemented
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getBranchReviewStatuses(id: idT) {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 70));
  return [];
  // if (id == 1) return ["accept", "reject", "accept"];
  // else if (id == 2) return ["accept", "open", "open"];
  // else return ["accept", "accept", "accept"];
}
