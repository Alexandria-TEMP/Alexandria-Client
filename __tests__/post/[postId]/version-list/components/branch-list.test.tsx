import { getBranchData } from "@/lib/api-calls/branch-api";
import BranchList from "@/post/[postId]/(main-post-view)/version-list/components/branch-list";
import { expect, describe, it } from "@jest/globals";
import { render } from "@testing-library/react";
import { dummyBranches } from "~/__tests__/__utils__/dummys";
import BranchCard from "@/post/[postId]/(main-post-view)/version-list/components/branch-card";

jest.mock("@/lib/api-calls/branch-api");
jest.mock(
  "@/post/[postId]/(main-post-view)/version-list/components/branch-card",
);

describe("BranchList", () => {
  (getBranchData as jest.Mock).mockResolvedValue(dummyBranches["open"]);
  (BranchCard as jest.Mock).mockReturnValue(<p>Card</p>);

  it("matches snapshot", () => {
    const { container } = render(
      <BranchList postId={1} ids={["0", "1", "2", "3", "4"]} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("grid matches snapshot", () => {
    const { container } = render(
      <BranchList grid postId={1} ids={["0", "1", "2", "3", "4"]} />,
    );
    expect(container).toMatchSnapshot();
  });
});
