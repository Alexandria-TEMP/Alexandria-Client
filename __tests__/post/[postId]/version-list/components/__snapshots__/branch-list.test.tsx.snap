// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BranchList grid matches snapshot 1`] = `
<div>
  <div
    class="gap-5 grid grid-cols-4"
  >
    <p>
      Card
    </p>
    <p>
      Card
    </p>
    <p>
      Card
    </p>
  </div>
</div>
`;

exports[`BranchList matches snapshot 1`] = `
<div>
  <div
    class="gap-5 flex flex-col"
  >
    <p>
      Card
    </p>
    <p>
      Card
    </p>
    <p>
      Card
    </p>
  </div>
</div>
`;

exports[`BranchList matches snapshot with no branches 1`] = `
<div>
  <h3
    class="text-foreground-500"
  >
    Looks like there are no versions here
  </h3>
</div>
`;
